### DISCUSSION TITLE
Poetry doesn't use my own virtual environment

### DISCUSSION THREAD

**Discussion by NielsRogge** [python-poetry/poetry#7767] [General]
I'd like to create a project with poetry. Here's my workflow (all done inside a folder in which I have a pyproject.toml file defined):
```
python -m venv env
source env/bin/activate
pip install poetry
poetry install
```
However, it seems that poetry installs all dependencies defined in the pyproject.toml not in the "env" virtual environment but in its own virtual environment. The package in my case is called "fondant", and when wanting to run it I get:
```
(env) (base) nielsrogge@Nielss-MacBook-Pro fondant % pip show fondant
WARNING: Package(s) not found: fondant
```
Also, neither of all dependencies of fondant (defined in the pyproject.toml file) are present in the virtual environment.
On the website is stated that:

> Poetry will detect and respect an existing virtual environment that has been externally activated. This is a powerful mechanism that is intended to be an alternative to Poetry’s built-in, simplified environment management.

Shouldn't it automatically recognize the virtual environment and install packages there?

I'm running everything in VSCode, on a Macbook Pro with MacOS Monterey 12.5.1.

Any help is appreciated!

Update:

when activating the virtual environment that Poetry created by itself by running `poetry shell`, I can import the packages and see all installed dependencies being installed there:
```
poetry shell
pip show fondant
```
But this is not really in line with what is stated on the website.

**Comment by clintonroy** [comment]
Do not install poetry into the same env as your project.

**Reply by NielsRogge** [reply]
Are you referring to pip install poetry or poetry install? Or both?

**Comment by sbrunner** [comment]
Poetry creates he's own virtual environment, then he doesn't install anything in your virtual environment...

**Reply by NielsRogge** [reply]
Yes I know, but on the website is stated that if you have your own virtual environment, Poetry will recognize it automatically and install packages there. But that doesn't seem to work for me

**Reply by ZelphirKaltstahl** [reply]
But the problem is of course, that Poetry does not offer a way to tell it _which_ virtual environment to install into, but prefers to instead create some rando folder somewhere with unpredictable name ...

**Reply by radoering** [reply]
> Poetry does not offer a way to tell it _which_ virtual environment to install into

If you do not want Poetry to create a venv, you can just activate a venv and Poetry will use it. I just tested it again and it works on my machine.

**Reply by ZelphirKaltstahl** [reply]
Yes, that works, and I have used it like that for a long time. However, at least until recently the Poetry documentation website itself stated, that one should not install poetry in the same environment that it manages. Currently I cannot find that advise on the documentation website any longer, so hopefully it is no longer the case (?). I also see some new config values like use active python. Perhaps the poetry team has realized this is an important use case and made the necessary changes.

**Reply by Secrus** [reply]
This is still the case that you shouldn't install Poetry into your project environment.

**Reply by ZelphirKaltstahl** [reply]
Could you link to that statement in the docs? 'cause I couldn't find it any longer. That would mean, that this way of installing dependencies goes against recommendation of the project maintainers. Also, could you link to an explanation why it is not recommended? Is it due to the dependencies of Poetry itself, that could conflict with whatever one wants to install for the project? And lastly, what is then the recommended way to install dependencies into a self-defined venv, according to documentation and maintainers?

**Reply by Secrus** [reply]
It's right there on the first page of the documentation: https://python-poetry.org/docs/#installation

**Reply by ZelphirKaltstahl** [reply]
Ah, I only looked in "managing environments", where it would be good to have it as well. So, now what about my other questions?

**Reply by Secrus** [reply]
> Ah, I only looked in "managing environments", where it would be good to have it as well.

Pull requests are welcome.

> So, now what about my other questions?

Create env, activate it, `poetry install`.

**Comment by ifeeltiredboss** [comment]
Why do you want to nest the environments?
`poetry install` will create a virtual environment and then install dependencies inside.

Also, you can disable this behavior so that `poetry install` won't create any environments, just install dependencies with
```
poetry config virtualenvs.create false
```

**Reply by NielsRogge** [reply]
Thanks, on the website is stated that Poetry will automatically recognize a virtual environment if it already exists, but that doesn't seem to work for me, hence my question:

> To achieve this, it will first check if it’s currently running inside a virtual environment. If it is, it will use it directly without creating a new one. But if it’s not, it will use one that it has already created or create a brand new one for you.

Source: https://python-poetry.org/docs/managing-environments/

**Comment by kharigardner** [comment]
Did you ever find a fix for this? All of a sudden, I've ran into some venv issues and have created a new one, but now poetry refuses to detect it and keeps creating new ones evertime I attempt to start my project.

**Reply by NielsRogge** [reply]
Hi,

So I would not recommend to create your own virtual environment, and instead let Poetry handle its own. 

Once you install a certain project using poetry (by running `poetry install` from the root of a repository), you can run `poetry shell` to activate it.

**Comment by Denilson-Semedo** [comment]


**Reply by earonesty** [reply]
nope, this just won't insteall the current package.   `poetry config virtualenvs.create false` is the correct response

**Comment by tjvananne** [comment]
TLDR: try naming your venv `.venv`, per the docs.

---

I'm on poetry 1.8.2. I'm not sure if this has changed since this original post, but the [docs](https://python-poetry.org/docs/configuration/#virtualenvscreate) seem to indicate the venv should be named `.venv`. I'm unsure if that actually matters:

> Create a new virtual environment if one doesn’t already exist.
> 
> If set to `false`, Poetry will not create a new virtual environment. If it detects an already enabled virtual environment or an existing one in `{cache-dir}/virtualenvs` or `{project-dir}/.venv` it will install dependencies into them, otherwise it will install dependencies into the systems python environment.

The following steps worked for me:

```bash
poetry config virtualenvs.create false --local
python3 -m venv .venv
poetry install
```

**Reply by dimbleby** [reply]
> `poetry config virtualenvs.create false`

this is [almost always bad advice](https://github.com/python-poetry/poetry/issues/6398)

**Reply by abelshanoj** [reply]
yeah it worked for me,naming the virtual env .venv does the trick..!

**Reply by earonesty** [reply]
venv without a "." seems to work.   "env" doesn't work.

**Reply by RamonGiovane** [reply]
> > `poetry config virtualenvs.create false`
> 
> this is [almost always bad advice](https://github.com/python-poetry/poetry/issues/6398)

poetry ignoring `env use` command is worse.
pick your poison, anyways

**Comment by danbailo** [comment]
Use this config after install `poetry`

```
pip install poetry
poetry config virtualenvs.create false --local
poetry config virtualenvs.prefer-active-python true --local
```

Basically this will say to poetry don't create a venv and will preffer your active python, in this case, the python from venv.

You can see all poetry config calling `$ poetry config --list`
```
cache-dir = "/home/daniel/.cache/pypoetry"
experimental.system-git-client = false
installer.max-workers = null
installer.modern-installation = true
installer.no-binary = null
installer.parallel = true
keyring.enabled = true
solver.lazy-wheel = true
virtualenvs.create = false
virtualenvs.in-project = null
virtualenvs.options.always-copy = false
virtualenvs.options.no-pip = false
virtualenvs.options.no-setuptools = false
virtualenvs.options.system-site-packages = false
virtualenvs.path = "{cache-dir}/virtualenvs"  # /home/daniel/.cache/pypoetry/virtualenvs
virtualenvs.prefer-active-python = true
virtualenvs.prompt = "{project_name}-py{python_version}"
warnings.export = true

```

**Comment by codeninja** [comment]
It is likely that VS Code is the culprit for you. If you have set your python interpreter to another project it may refuse to change the ENV as the python version that gets executed will always point to that other project in terminal even if you specify the path manually to `poetry env use`.

In my instance I had to bump out to the main ubuntu instance where `which python` properly reported `/usr/bin/python -> /usr/bin/python3.12` and then issue `poetry env use python` which properly created my `.venv`. 

From there I could `source .venv/bin/deactivate` which then allows my `which python` from within vscode to report `/usr/bin/python`... I can then `poetry env use` to use the venv in the project folder.

Also, verify that you're saving your env to `.venv` I've found that to be magical to resolve some issues.

PITA...  

![image](https://github.com/user-attachments/assets/ac5e2f54-fdce-4f2a-a0f1-7816c7b6647e)

**Comment by Christopher-Karvelis** [comment]
what worked for me:

- set the following poetry configuration:

`poetry config virtualenvs.prefer-active-python true`

- create and activate a venv and make poetry use it:

```
python -m venv env
source env/bin/activate
pip install poetry
poetry env use system
poetry install
```

**Comment by wooyek** [comment]
For anyone that has the same problem, please check if you don't have conda activated. I had a short brush with conda environment and forgot to deactivate it, assuming that activation is short lived and for the current terminal session only. 

Check if `CONDA_DEFAULT_ENV` is set in your current terminal session.

Since the ‎EnvManager.get checks it as well beside plain virtual environment, the check below failed to use `VIRTUAL_ENV` and failed to system python from pipx env. 

https://github.com/python-poetry/poetry/blob/341071a3f01379bf20682abf272c8fe8d10734e5/src/poetry/utils/env/env_manager.py#L224

**Comment by makslevental** [comment]
as of jul 24 2025 it's `poetry config virtualenvs.use-poetry-python false` instead of `poetry config virtualenvs.prefer-active-python true`
